trin-core/src/portalnet/overlay/service.rs

```diff
@@ -0,0 +1,741 @@
+        match request.direction {
+            RequestDirection::Incoming { source } => {
+                let response = self
+                    .handle_request(request.request.clone(), source.clone())
+                    .await;
+                if let Some(responder) = request.responder {
+                    let _ = responder.send(response);
+                }
+                match request.request {
+                    Request::Ping(ping) => self.process_ping(ping, source).await,
+                    _ => {}
+                }
+            }
+            RequestDirection::Outgoing { destination } => {
+                let response = self
+                    .send_talk_req(request.request, destination.clone())
+                    .await;
+                if let Some(responder) = request.responder {
+                    let _ = responder.send(response.clone());
+                }
+                match response {
+                    Ok(response) => {
+                        self.process_response(response, destination).await;
+                    }
+                    Err(error) => {
+                        debug!("Request failed with error {}", error);
+                        self.process_request_failure(destination).await
+                    }
+                }
+            }
+        }
+    }
+
+    /// Attempts to build a response for a request.
+    async fn handle_request(
+        &mut self,
+        request: Request,
+        source: NodeId,
+    ) -> Result<Response, String> {
+        match request {
+            Request::Ping(ping) => Ok(Response::Pong(self.handle_ping(ping, source).await?)),
+            Request::FindNodes(find_nodes) => {
+                Ok(Response::Nodes(self.handle_find_nodes(find_nodes).await?))
+            }
+            Request::FindContent(find_content) => Ok(Response::FoundContent(
+                self.handle_find_content(find_content).await?,
+            )),
+        }
+    }
+
+    /// Attempts to build a Pong response for a Ping request.
+    async fn handle_ping(&self, request: Ping, source: NodeId) -> Result<Pong, String> {
+        debug!(
+            "Got overlay ping request {:?} from node. Node: {}",
+            request, source
+        );
+        let enr_seq = self.local_enr().await.seq();
+        let data_radius = self.data_radius().await;
+        Ok(Pong {
+            enr_seq,
+            data_radius,
+        })
+    }
+
+    /// Attempts to build a Nodes repsonse for a FindNodes request.
+    async fn handle_find_nodes(&self, request: FindNodes) -> Result<Nodes, String> {
+        let distances64: Vec<u64> = request.distances.iter().map(|x| (*x).into()).collect();
+        let enrs = self.nodes_by_distance(distances64).await;
+        // from spec: total = The total number of Nodes response messages being sent.
+        // TODO: support returning multiple messages
+        Ok(Nodes { total: 1, enrs })
```

Comment:
What is the error case in some of these functions? `"Attempts to build"` but it doesn't look like it fails? And if it does fail, it should return an error. Or it should just not return a `Result` in the first place